src/transformers/generation/tf_logits_process.py

```diff
@@ -585,3 +586,137 @@ def _force_token(generation_idx):
+            tf.math.count_nonzero(((input_ids[:, self.begin_index : cur_len + 1] != self.eos_token_id)), axis=-1) >= 1
+        )
+        last_was_timestamps = not_eos & (input_ids[:, cur_len - 1] >= self.timestamp_begin)
+
+        not_eos = (
+            tf.math.count_nonzero((input_ids[:, self.begin_index : cur_len + 1] != self.eos_token_id), axis=-1) < 2
+        )
+        penultimate_was_timestamp = not_eos | (input_ids[:, cur_len - 2] >= self.timestamp_begin)
+
+        # timestamps have to appear in pairs, except directly before eos_token; mask logits accordingly
+        # for k in range(input_ids.shape[0]):
+        #     seq = [t for t in input_ids[k, self.begin_index :].tolist()]
+        #     last_was_timestamp = len(seq) >= 1 and seq[-1] >= self.timestamp_begin
+        #     penultimate_was_timestamp = len(seq) < 2 or seq[-2] >= self.timestamp_begin
+        #     if last_was_timestamp:
+        #          if penultimate_was_timestamp:  # has to be non-timestamp
+        #               scores[k, self.timestamp_begin :] = -float("inf")
+        #          else:  # cannot be normal text tokens
+        #               scores[k, : self.eos_token_id] = -float("inf")
+
+        def _update_slices(batch_idx_to_modify, idx0, idx1):
+            batch_idx_to_modify = tf.where(batch_idx_to_modify)
+            new_scores = tf.identity(scores)
+            if batch_idx_to_modify.shape[0] is not None:
+                if batch_idx_to_modify.shape[0] > 0:
+                    batches = batch_idx_to_modify.shape[0]
+                    mod = tf.tile(range(idx0, idx1), [batches])
+                    batch = tf.repeat(batch_idx_to_modify, (idx1 - idx0))
+                    final_idx = tf.stack((tf.cast(batch, mod.dtype), mod), axis=1)
+                    updates = tf.ones(((idx1 - idx0) * batches,), dtype=scores.dtype) * -float("inf")
+                    new_scores = tf.tensor_scatter_nd_update(new_scores, final_idx, updates)
+            return new_scores
+
+        max_idx = scores.shape[1]
+        indices_1 = last_was_timestamps & penultimate_was_timestamp
+        indices_2 = last_was_timestamps & tf.math.logical_not(penultimate_was_timestamp)
+
+        scores = _update_slices(indices_1, self.timestamp_begin, max_idx)
+        scores = _update_slices(indices_2, 0, self.eos_token_id)
+
+        # if input_ids.shape[1] == self.begin_index and self.max_initial_timestamp_index is not None:
+        #     last_allowed = self.timestamp_begin + self.max_initial_timestamp_index
+        #     scores[:, last_allowed + 1 :] = -float("inf")
+
+        scores = tf.cond(
+            tf.equal(cur_len, self.begin_index) and self.max_initial_timestamp_index is not None,
+            lambda: _update_slices(
+                tf.ones((scores.shape[0],), dtype=tf.bool),
+                self.timestamp_begin + self.max_initial_timestamp_index + 1,
+                max_idx,
+            ),
+            lambda: tf.identity(scores),
+        )
```

Comment:
1 - same masking strategy as above can be applied, if needed :)
2 - `cur_len`, `self.begin_index`, and `self.max_initial_timestamp_index` are not tensors (if I got it right), so a `tf.cond` should not be needed